Invalid offset when reading record from an older version

**Describe the bug**

```
java.lang.RuntimeException: Failed to replay batch at 'LoggedEvent [position=405952032, key=-1, timestamp=1718125492239, sourceEventPosition=405952031] RecordMetadata{recordType=EVENT, valueType=CHECKPOINT, intent=CREATED, rejectionType=INVALID_ARGUMENT, authorization=UNKNOWN}'
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayNextEvent$4(ReplayStateMachine.java:170) ~[zeebe-stream-platform-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) ~[zeebe-scheduler-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:96) ~[zeebe-scheduler-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47) [zeebe-scheduler-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:122) [zeebe-scheduler-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.3.12.jar:8.3.12]
Caused by: java.lang.IllegalArgumentException: invalid offset: -65442
	at org.agrona.concurrent.UnsafeBuffer.boundsCheckWrap(UnsafeBuffer.java:662) ~[agrona-1.19.2.jar:1.19.2]
	at org.agrona.concurrent.UnsafeBuffer.wrap(UnsafeBuffer.java:286) ~[agrona-1.19.2.jar:1.19.2]
	at io.camunda.zeebe.msgpack.spec.MsgPackReader.wrap(MsgPackReader.java:49) ~[zeebe-msgpack-core-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:40) ~[zeebe-msgpack-value-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.logstreams.impl.log.LoggedEventImpl.readValue(LoggedEventImpl.java:106) ~[zeebe-logstreams-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.stream.impl.records.RecordValues.readRecordValue(RecordValues.java:34) ~[zeebe-stream-platform-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.readRecordValue(ReplayStateMachine.java:312) ~[zeebe-stream-platform-8.3.12.jar:8.3.12]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.replayEvent(ReplayStateMachine.java:229) ~[zeebe-stream-platform-8.3.12.jar:8.3.12]
```

Following events happened before this error.

1. Cluster is at version 8.2.15
2. Cluster takes a backup
3. Cluster restores from this backup
4. After sometime, it was updated to 8.2.28. This is when we first see this error.

Root cause analysis:

To allow larger records, we had changes the metadataLength field in `LogEntryDescriptor` from short to int. https://github.com/camunda/camunda/pull/15991 The assumption was that since we use little endian, [changing the read from getShort() to getInt()](https://github.com/camunda/camunda/pull/15991/commits/160ae30b9cf28879df92d298ce1d700b86552c5c) will return the same value for the old version which used a short. This is mostly correct. But in this case, there were garbage values in the unused higher bytes. As a result `metadataLength` read as int returned invalid value. 

Why was garbage value here? The journal was truncated immediately after restore (this is expected.). During truncation we only mark the record as invalid by writing zeros to the header. We do not zero the whole record. The new record was written to the same location as the previous record. Since we did not zeroed it, there might be garbage values in the previously unused bytes. The version which writes this record was also 8.2.15 which writes the metadata as short. So it do not overwrite these unused bytes. 

After the broker is upgraded to a version that used integer for `metadataLength`, it reads these garbage bytes resulting in invalid length.

**To Reproduce**

Restore from backup in a version from before the above bug fix. After that, upgrade the cluster to a version with the above bug fix, before it has a chance to take a snapshot and compact. 

Related to https://camunda.slack.com/archives/C077X0MKYQ5

--- a/zeebe/logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogEntryDescriptor.java
+++ b/zeebe/logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogEntryDescriptor.java
@@ -51,6 +51,7 @@ public final class LogEntryDescriptor {
 
   public static final long KEY_NULL_VALUE = -1;
 
+  // When VERSION is incremented, also update the version check in getMetadataLength
   private static final short VERSION = 1;
   private static final int VERSION_OFFSET;
 
@@ -190,7 +191,12 @@ public final class LogEntryDescriptor {
   }
 
   public static int getMetadataLength(final DirectBuffer buffer, final int offset) {
-    return buffer.getInt(metadataLengthOffset(offset), Protocol.ENDIANNESS);
+    if (getVersion(buffer, offset) == 1) {
+      return buffer.getInt(metadataLengthOffset(offset), Protocol.ENDIANNESS);
+    } else {
+      // previous versions never set a version, so the version could be a garbage value or 0.
+      return buffer.getShort(metadataLengthOffset(offset), Protocol.ENDIANNESS);
+    }
   }
 
   public static void setMetadataLength(

--- a/zeebe/logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogEntryDescriptorTest.java
+++ b/zeebe/logstreams/src/test/java/io/camunda/zeebe/logstreams/impl/log/LogEntryDescriptorTest.java
@@ -7,14 +7,16 @@
  */
 package io.camunda.zeebe.logstreams.impl.log;
 
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.camunda.zeebe.protocol.Protocol;
+import java.util.Arrays;
 import org.agrona.concurrent.UnsafeBuffer;
-import org.assertj.core.api.Assertions;
 import org.junit.jupiter.api.Test;
 
-public class LogEntryDescriptorTest {
-
+final class LogEntryDescriptorTest {
   @Test
-  public void shouldBeNonProcessedAsDefault() {
+  void shouldBeNonProcessedAsDefault() {
     // given
     final var buffer = new UnsafeBuffer(new byte[128]);
 
@@ -22,11 +24,11 @@ public class LogEntryDescriptorTest {
     final boolean processed = LogEntryDescriptor.shouldSkipProcessing(buffer, 0);
 
     // then
-    Assertions.assertThat(processed).isFalse();
+    assertThat(processed).isFalse();
   }
 
   @Test
-  public void shouldMarkAsProcessed() {
+  void shouldMarkAsProcessed() {
     // given
     final var buffer = new UnsafeBuffer(new byte[128]);
 
@@ -34,6 +36,64 @@ public class LogEntryDescriptorTest {
     LogEntryDescriptor.skipProcessing(buffer, 0);
 
     // then
-    Assertions.assertThat(LogEntryDescriptor.shouldSkipProcessing(buffer, 0)).isTrue();
+    assertThat(LogEntryDescriptor.shouldSkipProcessing(buffer, 0)).isTrue();
+  }
+
+  @Test
+  void shouldReadShortMetadataLength() {
+    // given
+    final var buffer = new UnsafeBuffer(new byte[128]);
+
+    // when
+    LogEntryDescriptor.setVersion(buffer, 0);
+    LogEntryDescriptor.setMetadataLength(buffer, 0, 34);
+
+    // then
+    assertThat(LogEntryDescriptor.getMetadataLength(buffer, 0)).isEqualTo(34);
+  }
+
+  @Test
+  void shouldReadLargeMetadataLength() {
+    // given
+    final var buffer = new UnsafeBuffer(new byte[128]);
+
+    // when
+    LogEntryDescriptor.setVersion(buffer, 0);
+    LogEntryDescriptor.setMetadataLength(buffer, 0, Short.MAX_VALUE + 10);
+
+    // then
+    assertThat(LogEntryDescriptor.getMetadataLength(buffer, 0)).isEqualTo(Short.MAX_VALUE + 10);
+  }
+
+  @Test
+  void shouldReadMetadataLengthFromOldVersion() {
+    // given
+    final byte[] byteArray = new byte[128];
+    Arrays.fill(byteArray, (byte) 8);
+    final var buffer = new UnsafeBuffer(byteArray);
+
+    // when
+    // set version 0
+    buffer.putShort(LogEntryDescriptor.versionOffset(0), (short) 0, Protocol.ENDIANNESS);
+    LogEntryDescriptor.setMetadataLength(buffer, 0, 34);
+
+    // then
+    assertThat(LogEntryDescriptor.getMetadataLength(buffer, 0)).isEqualTo(34);
+  }
+
+  @Test
+  void shouldReadMetadataLengthFromOldVersionWhenVersionNotSet() {
+    // given
+    final byte[] byteArray = new byte[128];
+    Arrays.fill(byteArray, (byte) 8);
+    final var buffer = new UnsafeBuffer(byteArray);
+
+    // when
+    // set garbage value for version
+    buffer.putShort(LogEntryDescriptor.versionOffset(0), (short) 123, Protocol.ENDIANNESS);
+    LogEntryDescriptor.setMetadataLength(buffer, 0, 34);
+
+    // then
+    assertThat(LogEntryDescriptor.getMetadataLength(buffer, 0)).isEqualTo(34);
   }
 }

